Sources/YUL/AST.swift

```diff
@@ -0,0 +1,342 @@
+    self.expression = expression
+    self.block = block
+  }
+
+  public var description: String {
+    return "if \(expression.description) \(self.block)"
+  }
+}
+
+public struct Switch: CustomStringConvertible {
+  public let expression: Expression
+  public let cases: [(Literal, Block)]
+  public let `default`: Block?
+
+  public init(_ expression: Expression, cases: [(Literal, Block)], `default`: Block? = nil) {
+    self.expression = expression
+    self.cases = cases
+    self.default = `default`
+  }
+
+  public init(_ expression: Expression, `default`: Block? = nil) {
+    self.init(expression, cases: [], default: `default`)
+  }
+
+  public var description: String {
+    let cases = self.cases.map { (lit, block) in
+      return "case \(lit) \(block)"
+      }.joined(separator: "\n")
+
+    let `default` = self.default != nil ? "\ndefault \(self.default!)" : ""
+
+    return """
+    switch \(self.expression)
+    \(cases)\(`default`)
+    """
+  }
+}
+
+public struct ForLoop: CustomStringConvertible {
+  public let initialize: Block
+  public let condition: Expression
+  public let step: Block
+  public let body: Block
+
+  public init(_ initialize: Block, _ condition: Expression, _ step: Block, _ body: Block) {
+    self.initialize = initialize
+    self.condition = condition
+    self.step = step
+    self.body = body
+  }
+
+  public var description: String {
+    return "for \(initialize) \(condition) \(step) \(body)"
+  }
+}
+
+public enum Statement: CustomStringConvertible {
+  case block(Block)
+  case functionDefinition(FunctionDefinition)
+  case `if`(If)
+  case expression(Expression)
+  case `switch`(Switch)
+  case `for`(ForLoop)
+  case `break`
+  case `continue`
+  case noop
+  case inline(String)
+
+  public var catchableSuccesses: [Expression] {
+    switch self {
+    case .if(let ifs):
+      return ifs.expression.catchableSuccesses
+    case .expression(let e):
+      return e.catchableSuccesses
+    case .switch(let sw):
+      return sw.expression.catchableSuccesses
+    default:
+      return []
+    }
+  }
+
+  public var description: String {
+    switch self {
+    case .block(let b):
+      return b.description
+    case .functionDefinition(let f):
+      return f.description
+    case .if(let ifs):
+      return ifs.description
+    case .expression(let e):
+      return e.description
+    case .switch(let sw):
+      return sw.description
+    case .`for`(let loop):
+      return loop.description
+    case .break:
+      return "break"
+    case .continue:
+      return "continue"
+    case .noop:
+      return ""
+    case .inline(let s):
+      return s
+    }
+  }
+}
+
+public enum Type: String, CustomStringConvertible {
+  case bool
+  case u8
+  case s8
+  case u32
+  case s32
+  case u64
+  case s64
+  case u128
+  case s128
+  case u256
+  case s256
+  case any
+
+  public var description: String {
+    return self.rawValue
+  }
+}
+
+public typealias TypedIdentifierList = [(Identifier, Type)]
```

Comment:
Why do we need to alias the list? Personally I think it'd be better if we typealias just the tuple as `TypedIdentifier` and when we can use it either in a list context or (presumably at some point) just on its own